### PR TITLE
Translate username in Notebook label and dashboard config user states

### LINKED ISSUE #316: [KFNBC]: Notebooks cannot be created with `kube:admin` user
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When trying to spin up a Notebook for my `kube:admin` username, I get a 404 error from the backend and nothing happens on the frontend.

All I saw in the UI when I clicked on the "Start Server" button was it go grey for a few moments and then re-enable. The console log said 404 & Request Failed... and the actual error was hidden, took some debugging to figure it out.

### Expected Behavior

For the notebook to be created or an error to be shown to the user (in this case, no error would have been expected as there was nothing I could do -- but the user shouldn't be left in the dark either way)

https://github.com/opendatahub-io/odh-dashboard/issues/281 looks to improve the error handling side of things

### Steps To Reproduce

1. Create & log in as `kube:admin`
2. Navigate into the KFNBC Spawner Page
3. Any settings as long as your user name has a `:` in it
4. Start Server

### Workaround (if any)

None -- usernames are often not configured by the user themselves.

### OpenShift Version

_No response_

### Openshift Version

_No response_

### What browsers are you seeing the problem on?

_No response_

### Open Data Hub Version

_No response_

### Relevant log output

```shell
// The error that was swallowed was:

  {
    reason: 'FieldValueInvalid',
    message: `Invalid value: "kube:admin": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')`,
    field: 'metadata.labels'
  }
```

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
Closes #316 
There is an issue to generate a userState entry with username equals to a `''`, I also fixed that in this PR.

### CODE REVIEW

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/src/pages/notebookController/SpawnerPage.tsx [hunk 1]
```diff
@@ -65,6 +71,18 @@ const SpawnerPage: React.FC<SpawnerPageProps> = React.memo(({ setStartModalShown
...
+  React.useEffect(() => {
+    if (translatedUsername) {
+      const newUserState = dashboardConfig?.spec.notebookControllerState?.find(
+        (state) => state.user === translatedUsername,
+      );
+      if (newUserState) {
+        setUserState(newUserState);
+      }
+    }
+  }, [dashboardConfig, translatedUsername]);
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/pages/notebookController/SpawnerPage.tsx]: Updated, thanks!

File: frontend/src/pages/notebookController/SpawnerPage.tsx [hunk 2]
```diff
@@ -65,6 +71,18 @@ const SpawnerPage: React.FC<SpawnerPageProps> = React.memo(({ setStartModalShown
...
+  React.useMemo(() => {
+    if (translatedUsername) {
+      const newUserState = dashboardConfig?.spec.notebookControllerState?.find(
+        (state) => state.user === translatedUsername,
+      );
+      if (newUserState) {
+        setUserState(newUserState);
+      }
+    }
+  }, [dashboardConfig, translatedUsername]);
```
Comment (andrewballantyne) [inline frontend/src/pages/notebookController/SpawnerPage.tsx:85]:
```suggestion
  const userState = React.useMemo(() => {
    if (translatedUsername) {
      const newUserState = dashboardConfig?.spec.notebookControllerState?.find(
        (state) => state.user === translatedUsername,
      );
      return newUserState;
    } else {
      return EMPTY_USER_STATE;
    }
  }, [dashboardConfig, translatedUsername]);
```

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
`useMemo` is just `useEffect` to save in a `useState`... you do still need to get the data out of a `useMemo`.

See the documentation https://reactjs.org/docs/hooks-reference.html#usememo

File: frontend/src/pages/notebookController/SpawnerPage.tsx [hunk 3]
```diff
@@ -65,6 +71,16 @@ const SpawnerPage: React.FC<SpawnerPageProps> = React.memo(({ setStartModalShown
...
   const [createInProgress, setCreateInProgress] = React.useState<boolean>(false);
+  const userState = React.useMemo(() => {
+    if (translatedUsername) {
+      const newUserState = dashboardConfig?.spec.notebookControllerState?.find(
+        (state) => state.user === translatedUsername,
+      );
+      return newUserState;
+    } else {
+      return EMPTY_USER_STATE;
+    }
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/pages/notebookController/SpawnerPage.tsx:82]: LOL, I didn't even notice this, thanks for the fix!

**Review by andrewballantyne — APPROVED** [review]
heh, sorry about the typo -- I'll try to write them in my IDE before giving them here

**Review by andrewballantyne — COMMENTED** [review]
Ah you got a linter failure... it wants to reorganize the import line breaks. Weird that wasn't there before 🤔 run `npm run lint` in the frontend folder and fix the errors.

**Comment by samuelvl** [comment]
I could spawn a notebook with a normal user and with `kube:admin`:

```
$ oc get pods -l opendatahub.io/user=kube-3aadmin
NAME                        READY   STATUS    RESTARTS   AGE
jupyter-nb-kube-3aadmin-0   2/2     Running   0          75s
```

**Review by samuelvl — APPROVED** [review]

**Pull request merged**